doc/tunnel/SAI-Proposal-ImprovedP2PTunnels.md

```diff
@@ -0,0 +1,61 @@
+# Introduction
+As of SAI version 1.5 the “tunnel” has a p2mp connotation. It holds the VTEP SIP whereas there is no DIP. The DIP is specified as part of the FDB entry or as part of Next Hop entry.
+
+It is proposed to add the DIP as part of the `sai_tunnel_attr_t` structure as an optional parameter.
+
+The following are the motivations for introducing a DIP in the tunnel structure and to model it as a p2p entity.
+
+- Support Head End Replication as the method for handling BUM traffic.
```

Comment:
There is an API to configure the flood vector for BUM traffic:

```
// add tunnel endpoints 1.1.1.2, 1.1.1.3, 1.1.1.4 to flood vector
// vlan, tunnel, tunnel_bridge_port objects are pre-created

sai_object_id_t l2mc_group;
sai_object_id_t member2, member3, member4;

sai_attribute_t attr;
vector<sai_attribute_t> attrs;

// create L2MC group

sai_l2mc_group_api->cerate_l2mc_group(&l2mc_group, g_switch_id, attrs.size(), attrs.data());

// create L2MC group members

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = tunnel_bridge_port;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 0x01010102;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member2, g_switch_id, attrs.size(), attrs.data());

attrs.clear();

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = tunnel_bridge_port;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 0x01010103;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member3, g_switch_id, attrs.size(), attrs.data());

attrs.clear();

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = tunnel_bridge_port;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 0x01010104;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member4, g_switch_id, attrs.size(), attrs.data());

// set the L2MC group on a vlan

attr.id = SAI_VLAN_ATTR_UNKNOWN_UNICAST_FLOOD_CONTROL_TYPE;
attr.value.oid = SAI_VLAN_FLOOD_CONTROL_TYPE_COMBINED;

sai_vlan_api->set_vlan_attribute(vlan, &attr);

attr.id = SAI_VLAN_ATTR_UNKNOWN_UNICAST_FLOOD_GROUP;
attr.value.oid = l2mc_group;

sai_vlan_api->set_vlan_attribute(vlan, &attr);
```

Reply:
1. Is the output id per remote ip or 1 per source ip ? 
2. How would we achieve split horizon with this approach ? 
3. The API assumes a separate flood group for vxlan side members and a separate one for access side members.. How would a device with single flooding group handle the following 
  cases without burning up unnecessary flood groups ?
  - addition of first tunnel member after the access members have been added 
  - removal of last tunnel member after the access members have been added.

Reply:
It can also be done per remote IP
For the sake of conserving L2MC groups, the "read-modify-write" operation is also an option
```
/*****************************************************
 * Get flood vector on VLAN
 *****************************************************/

attr.id = SAI_VLAN_ATTR_UNKNOWN_UNICAST_FLOOD_GROUP;

sai_vlan_api->get_vlan_attribute(vlan, &attr);

sai_object_it l2mc_group = attr.value.oid;

/*****************************************************
 * create L2MC group members
 *****************************************************/
 
sai_object_id_t member2, member3, member4;


attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = bridge_port2;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 1.1.1.2;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member2, g_switch_id, attrs.size(), attrs.data());

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = bridge_port3;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 1.1.1.3;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member3, g_switch_id, attrs.size(), attrs.data());

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_GROUP_ID;
attr.value.oid = l2mc_group;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_OUTPUT_ID;
attr.value.oid = bridge_port4;
attrs.push_back(attr);

attr.id = SAI_L2MC_GROUP_MEMBER_ATTR_L2MC_ENDPOINT_IP;
attr.value.ipaddr.addr = 1.1.1.4;
attrs.push_back(attr);

sai_l2mc_group_api->cerate_l2mc_group_member(&member4, g_switch_id, attrs.size(), attrs.data());
```